### ISSUE TITLE
Show a couple of days in the coming weather

### ISSUE THREAD

**Issue by R0TEAM** [dovrosenberg/foundry-simple-weather#39]
Would be nice to know which weather is coming to say to the player "you see in the distance an dark cloud formation come ..." ;)
So he can prepere and planing routes/rest

**Comment by dovrosenberg** [comment]
I like the idea of a forecast, but right now the weather isn't actually connected to the day... A change of day is just one way that the weather can be triggered to change.

I also think at some point there may be a need to change weather mid-day (for ex. to be able to generate a much lower temperature at night in a desert), which will make this even more complex.

How would you see that interacting with a) the "regenerate" button and b) manually specifying weather?  If I hit regenerate should it regenerate the next few days?  Or move to the next "day" in the sequence?  Or something else?  If I manually pick weather, what happens to the forecast?

**Comment by SesuUisu** [comment]
For the tempeature at day and night, you could just generate it together (lowest and highest temp of the day). Having a hourly change would be too much for me. 

My suggestions for your questions:
a) with regeneration the next day could also change, but doesn't have to
b.1) I think there are situations, where I just want to have the normal weather as predicted before (like with magical changed weather)
b.2) but if it's just a situation like "adventure book says it's snowy" and I change it to it, the next day should go with the generator. 
b.3) there may be the situation, I would like to have the option to change the next day, but not the current one
b.4) If the PCs do not have the skills/knowledge to interprete the weather, I would like to hide the weather forecast

**Comment by Issacar13** [comment]
Could also make midnight and noon the trigger points for weather changes.

Also, having the option to hard set an upcoming day so that the algorithm chooses the weather you selected instead of generating a new one.

Also weather prediction doesn't have to be 100% accurate (like in reality).

**Comment by loofou** [comment]
Not to hijack this thread, but a way to set when the weather updates would be great. For example the Twilight 2000 4e system runs on four "shifts" in the day (Night, Morning, Day, Evening). Each should update the weather in a consistent and smart way.

A weather forecast would of course also be very helpful. One way to set this up is to use the stored information in Simple Calendar to load the weather information. This way weather could be generated for a number of days in advance and the module then loads the data each day from the calendar. If it doesn't exist, it simply generates a new one.

**Comment by dovrosenberg** [comment]
> Not to hijack this thread, but a way to set when the weather updates would be great. For example the Twilight 2000 4e system runs on four "shifts" in the day (Night, Morning, Day, Evening). Each should update the weather in a consistent and smart way.
> 

I'm adding the multiple times/day concept to another feature request.

**Comment by dovrosenberg** [comment]
OK - I think I have an approach here.  Would like comments @loofou  @R0TEAM @SesuUisu @Issacar13 before I start work on it.  I'm going to defer the multiple changes/day part for later, as that's going to be a lot more work because of how weather generation works today.

The basic idea for weather forecasts:
1. A setting to determine whether forecasts are created
2. A separate setting for whether or not forecasts are included in the chat (so the GM can see the forecast and decide how to reply if players ask vs. automatically providing it to players)
3. A setting for how many days of forecast are created
4. A setting for how accurate forecasts are (stated as a probability that the next day is correct).  I suspect values between 90% and 100% will prove most useful but will leave it open for whatever people want to try.
5. When forecasts are enabled, every time the weather changes it will generate the next N days of weather using the same algorithm it does today.  
6. It will display the forecast as ONLY a) a temperature, and b) a "sunny-ness" scale (clear, partly cloudy, fully cloudy, precipitation).  Additionally, the temperature shown will be the "average" used in the calculation, so there will still be some randomness around it when the day actually arrives.  I think this combination of random temperature and only a general description will better "smooth" the changes and make it less "binary" if the forecast was right or not.  This will be shown in the "today's weather" box, as well as put in the chat (based on setting)
7.  It will (probably based on another setting) store this in Simple Calendar notes on the future days, but note that it's just a forecast to avoid confusion.
8. When a day arrives that has been forecast, it will roll a probability.  If it falls inside the "forecast reliability" setting, it will a) just use the previously forecast weather and b) then regenerate one extra day at the end of the prior forecast.  If it's outisde (ex. in the 10% chance when reliability set to 90%) it will instead regenerate the entire forecast and force that a new weather is picked for the following day (otherwise the effective reliability will be higher than is set).  In some cases, it may still look like the next day was the same as forecast because some weather types with different descriptions will have the same "forecast description" (see item 6) but they'll lead to different paths on the following days.  

NOTE: A potential issue here is if I generate weather on Day 1, then skip to Day 3, the weather fore Day 2 will still be a forecast, and when ultimately generated may not make a lot of sense leading into Day 3.  But if people are going to jump around like that, I'm not sure there's a whole lot I can do... they can always set it manually.

Manually setting the weather for a day will still trigger the forecast as if the system had picked that weather.


Thoughts?

**Comment by loofou** [comment]
Looks good to me. I think you should also make sure to not overwrite already determined weather notes with forecasts, should someone want to jump _backwards_ in time (for a flashback or something). Unless the user manually recreates the weather for the day of course.

**Comment by SesuUisu** [comment]
For the forecast, as a GM I need a 100% probability, but for my players I think something between 0 and 90% depending on their skill level would be nice to have. I have no idea, if different outputs are feasible to do. But having something correct on hand would be nice, while my players should not know what the future will bring.

**Comment by dovrosenberg** [comment]
> Looks good to me. I think you should also make sure to not overwrite already determined weather notes with forecasts, should someone want to jump _backwards_ in time (for a flashback or something). Unless the user manually recreates the weather for the day of course.

Agree.  Once a day has weather recorded, that's it (unless you force a regenerate).  I'll figure out a way to work the actual ones into the forecast somehow if you're looking at a forecast that includes a day that already was generated.

**Comment by dovrosenberg** [comment]
> For the forecast, as a GM I need a 100% probability, but for my players I think something between 0 and 90% depending on their skill level would be nice to have. I have no idea, if different outputs are feasible to do. But having something correct on hand would be nice, while my players should not know what the future will bring.

Oooh.  Interesting.  What if there are 2 modes:
1) only the GM can see the forecast and it's 100% accurate
2) it displays to players and is probabilistic.  

I think there are two many systems and variables to have the player forecast automatically tie to anything that makes sense - instead it's available for people who want to give a sense of forecast without hassle.  For others, it would be up to the GM to say how the players would interpret the "actual" that they can see.

Would number 1 work for you @SesuUisu?  If this works, I'd probably build this version first and then add version 2.

**Comment by SesuUisu** [comment]
> > For the forecast, as a GM I need a 100% probability, but for my players I think something between 0 and 90% depending on their skill level would be nice to have. I have no idea, if different outputs are feasible to do. But having something correct on hand would be nice, while my players should not know what the future will bring.
> 
> Oooh.  Interesting.  What if there are 2 modes:
> 1) only the GM can see the forecast and it's 100% accurate
> 2) it displays to players and is probabilistic.  
> 
> I think there are two many systems and variables to have the player forecast automatically tie to anything that makes sense - instead it's available for people who want to give a sense of forecast without hassle.  For others, it would be up to the GM to say how the players would interpret the "actual" that they can see.
> 
> Would number 1 work for you @SesuUisu?  If this works, I'd probably build this version first and then add version 2.

Sounds good to me!
If I have any idea how the system/skill issue could be addressed, I will let you know.

**Comment by dovrosenberg** [comment]
Here's what I've got... A simple display of just the weather "symbols" with a tooltip that shows you the info for that day:

![image](https://github.com/user-attachments/assets/659f5758-2ea8-4175-8c9b-53523f2c9eb7)

Anything else this first version would need?

**Comment by bragginsgt93** [comment]
Good afternoon. I was wondering if you could give some insight to me. I have been using Simple weather along with Simple Calendar for quite some time now. Always updated, with the recent release 1.17.1 all the simple weather settings disappeared. Tried uninstalling/reinstalling. Loaded only that mod alone, reverted back to older versions. Nothing will bring back the settings. I am usrd 3.5 srd. Any help would be appreciated.

**Comment by dovrosenberg** [comment]
Closing this for now.  If anyone has new related requests, open a new issue.
